---
title: "`r params$report_title`"
output:
  html_document:
    toc: true
    toc_float: true
    theme: united
    code_folding: hide
params:
  report_title: "Check aneuploidy in RNAseq samples"
  report_author: "Anonymus"
  bam_folder: "../test_data/samples_list.csv"
  sample_metadata: "../test_data/samples_meta.csv"
  condition_column: "Label"
  condition_order: null
  num_cpu: 4
---

## Setup

### Initialize tools

```{r message=FALSE, warning=FALSE}
# Import packages

library(tidyr)
library(purrr)
library(ggplot2)
library(ggsci)
library(ggridges)
library(plotly)
library(pheatmap)
library(DT)
library(AneuFinder)
library(Repitools)
```

```{r}
Aneufinder(
  inputfolder=params$bam_folder, outputfolder='aneu_out',
  numCPU=params$num_cpu, method=c('edivisive', 'dnacopy', 'HMM')
)
```

```{r}
files <- list.files('/home/rstudio/local_files/data/aneu_out/MODELS/method-edivisive', full.names=TRUE)
hmms <- loadFromFiles(files)
```

```{r}
cl <- clusterByQuality(hmms)
heatmapGenomewide(cl$classification[[1]])
```

```{r}
getBreakpoints(hmms[[1]])
```

```{r}
plot(files[1], type='karyogram')
```

```{r}
plot(files[1], type='profile')
```

```{r fig.width=16, fig.height=9}
heatmapGenomewide(hmms, cluster=FALSE)
```

```{r fig.width=16, fig.height=9}
heatmapAneuploidies(hmms)
```

```{r}
karyotypeMeasures(files)
```

```{r fig.width=16, fig.height=9}
aneu_df <- heatmapAneuploidies(hmms, as.data.frame=TRUE)
head(aneu_df)
```

```{r}
aneuploidy_df <- hmms %>%
  purrr::map_dfr(function (x) annoGR2DF(x$segments), .id='FILE_LABEL') %>%
  mutate(
    SAMPLE_ID = basename(FILE_LABEL),
    SAMPLE_ID = gsub('\\.bam.*', '', SAMPLE_ID),
    SAMPLE_POS = as.numeric(factor(SAMPLE_ID)),
  )

head(aneuploidy_df)
```

```{r}
aneuploidy_df %>%
  ggplot(aes(
    xmin = start, xmax = end, ymin = SAMPLE_POS - 1 , ymax = SAMPLE_POS - 0.2,
    fill = state
  )) +
  geom_rect() +
  scale_fill_npg() +
  facet_wrap("chr", scale="free_x", nrow=3, strip.position="bottom") +
  theme(
    axis.text.x = element_blank(),
    axis.ticks = element_blank()
  )
```

```{r}
edivisive_scores <- files %>%
  setNames(., basename(.)) %>%
  map(karyotypeMeasures) %>%
  map_dfr(~bind_rows(
    tibble::rownames_to_column(.x$per.chromosome, "Chr"),
    mutate(.x$genomewide, Chr="GN")
  ), .id="SAMPLE_FILE") %>%
  mutate(
    METHOD = "edivisive"
  )
```

```{r}
edivisive_scores %>%
  filter(Chr == "GN") %>%
  ggplot(aes(x=SAMPLE_ID, y=Aneuploidy)) +
  geom_col(position="dodge") +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle = 30, vjust = 0.5, hjust=1)
  ) +
  labs(
    x="", title="Genome-wide aneuploidy scores"
  )
```

```{r}
dnacopy_matrix <- edivisive_scores %>%
  filter(Chr %in% c(seq(1, 21), "X", "Y")) %>%
  select(SAMPLE_LABEL, SAMPLE_ID, Aneuploidy, Chr) %>%
  pivot_wider(values_from=Aneuploidy, names_from=Chr)

dnacopy_matrix %>%
  tibble::column_to_rownames("SAMPLE_ID") %>%
  heatmaply(
    dendrogram = FALSE, hide_colorbar = TRUE,
    row_side_colors = "SAMPLE_LABEL", row_side_palette = pal_npg(),
    main = "Chromosome-wise aneuploidy scores according to edvisive method"
  )
```
